[New Tutorial]: Build Your Own translator app in 30 min or less (#663)
build-on-aws/content: tutorials/build-your-own-translator-app-in-less-30-min/index.md

> ---

Author: Thanks! :) I will review it and follow your comments

title: "Build Your Own translator app in 30 min or less"
description: "Use Amazon Translate, Amazon Comprehend, Amazon Lambda Function and Amazon Lex to Build an translator app in 30 min or less"

> tags:

Author: Changed to this: Use Amazon Translate, Amazon Comprehend, Amazon Lambda, Amazon Polly, and Amazon Lex to bring a translation application to life and test it in 30 minutes or less.

    - chatbot

authorGithubAlias: elizabethfuentes12
authorName: Elizabeth Fuentes
date: 2023-08-22
---

The Internet has connected people worldwide, but language remains a barrier to universal understanding. With over 7,000 languages spoken globally, communication is limited for many. Creating content and apps in multiple languages allows greater reach, and managing multiple languages enables communication with more people. 

Imagine creating your own translation app in just a few steps - one capable of identifying the input language, translating into multiple languages, and generating audio files with correct pronunciation. This is now possible by combining AWS services. 

While translation apps exist, building your own and getting hands-on experience is the best way to learn and gain new skills. Creating something yourself matters more than whether it already exists - the key is developing abilities through practice.

| Attributes                |                                   |
| ------------------- | -------------------------------------- |
| ✅ AWS Level        | Intermediate - 200                         |
| ⏱ Time to complete  | 30 minutes                             |

> | 💰 Cost to complete | [AWS Free Tear](https://aws.amazon.com/free/)      |

Reviewer: ```suggestion
| 💰 Cost to complete | [AWS Free Tier](https://aws.amazon.com/free/)      |
```

| 🧩 Prerequisites    | - [AWS Account](https://aws.amazon.com/resources/create-account/?sc_channel=el&sc_campaign=devopswave&sc_content=cicdcfnaws&sc_geo=mult&sc_country=mult&sc_outcome=acq) <br>-  [Foundational knowledge of Python](https://catalog.us-east-1.prod.workshops.aws/workshops/3d705026-9edc-40e8-b353-bdabb116c89c/)    |                           |
| 📢 Feedback            | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?    |
| ⏰ Last Updated     | 2023-08-22                             |

## What You Will Learn

- How to use Artificial Intelligence service to detect the dominant language in texts.
- How to use Artificial Intelligence service to translate text.

> - How to use Artificial Intelligence service to converts text into lifelike speech. 

Reviewer: ```suggestion
- How to use Artificial Intelligence service to convert text into lifelike speech. 
```

- How to create a Artificial Intelligence conversational interfaces bot to handle translation requests.

## Solution Overview

> In this blog you are going to create a translator chatbot app, with [Amazon Lex](https://aws.amazon.com/lex/) that will handle the frontend interaction with the user, and the backend will be in charge of an [AWS Lambda Function](https://aws.amazon.com/lambda/) with the [AWS SDK for Python library Boto3](https://aws.amazon.com/sdk-for-python/) code of the follow AWS services:

Reviewer: ```suggestion
In this blog you are going to create a translator chatbot app, with [Amazon Lex](https://aws.amazon.com/lex/) that will handle the frontend interaction with the user, and the backend will be in charge of an [AWS Lambda Function](https://aws.amazon.com/lambda/) with the [AWS SDK for Python library Boto3](https://aws.amazon.com/sdk-for-python/) code using the following AWS services:
```

-	[Amazon Comprehend](https://aws.amazon.com/comprehend/) in charge of detecting the language entered.
-	[Amazon Translate](https://aws.amazon.com/translate/) it will translate into the desired language.
-	[Amazon Polly](https://aws.amazon.com/polly/) it delivers the audio with the correct pronunciation. 

![Diagram translator chatbot app"](images/fig_1.png)
<div align="center">Fig 1. Diagram translator chatbot app. </div> 

> ### You will build it in five parts:

Reviewer: ```suggestion
### You will build it in seven parts:
```

In this part you will use the [Boto3 Polly client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/polly.html) to call the APIs [StartSpeechSynthesisTask](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/polly/client/start_speech_synthesis_task.html), and [GetSpeechSynthesisTask](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/polly/client/get_speech_synthesis_task.html) to retrieves information of SpeechSynthesisTask based on its TaskID, delivering status and a link to the [Amazon S3 Bucket](https://aws.amazon.com/s3/)containing the output of the task. 

![converts text into lifelike speech](images/fig_2.png)
<div align="center">Fig 2. Converts text into lifelike speech.</div>

### Parameters requires for Amazon Lex APIs:

|StartSpeechSynthesisTask Parameters | GetSpeechSynthesisTask Parameter|
| ------------------- | -------------------------------------- |

> | <ul><li>**OutputFormat(string)**: The format in which the returned output will be encoded. For audio stream, this will be mp3, ogg_vorbis, or pcm. For speech marks, this will be json. <li>**OutputS3BucketName (string)**: [Amazon S3](https://aws.amazon.com/s3/) bucket name to which the output file will be saved. <li>**Text (string)**: The input text to synthesize. <li>**Engine (string)**: Specifies the engine ( standard or neural) for Amazon Polly to use when processing input text for speech synthesis. <li>**VoiceId (string)**: [Voice ID](https://docs.aws.amazon.com/polly/latest/dg/voicelist.html) to use for the synthesis.</li></ul>|  <ul><li>**TaskId (string)**: The Amazon Polly generated identifier for a speech synthesis task.</li></ul>|

Author: Resolved: 
|StartSpeechSynthesisTask Parameters | GetSpeechSynthesisTask Parameter|
| ------------------- | -------------------------------------- |
| <ul><li>**OutputFormat(string)**: the value at which the result will be returned and it can be: *json*,*mp3*, *ogg_vorbis* or *pcm*. <li>**OutputS3BucketName (string)**: It's the [Amazon S3](https://aws.amazon.com/s3/) bucket name to which the output file will be saved. <li>**Text (string)**: The input text to synthesize. <li>**Engine (string)**: It specifies the engine (standard or neural) for Amazon Polly to use when processing input text for speech synthesis. <li>**VoiceId (string)**: [Voice ID](https://docs.aws.amazon.com/polly/latest/dg/voicelist.html) is used for the synthesis.</li></ul>|  <ul><li>**TaskId (string)**: This is the Amazon Polly generated identifier for a speech synthesis task.</li></ul>|
|||

|**Slot type**|language|Restrict to slot values|
|**Prompts**| What language do you want to translate? | What do you want to translate? Just type and I'll tell you the language... magic!|
|**Image result** | ![Slot language](images/fig_5.png) <div align="center">Fig 5. Slot language required for intent fulfillment. </div> |![Slot language](images/fig_6.png) <div align="center">Fig 6. Slot language required for intent fulfillment. </div> |    

>🚨**Important:** The order matters, make sure that `language` type is the first slot to be required. 

This bot will invoke a Lambda function as a [dialog code hook](https://docs.aws.amazon.com/lexv2/latest/dg/paths-code-hook.html), to validate user input and to fulfill the intent, for that select Use a Lambda function for initialization and validation (Fig. 7).

![Slot language](images/fig_7.png) <div align="center">Fig 7. Use a Lambda function for initialization and validation. </div> 

To finish creating the chatbot, press **Save intent** and then **Build** in the top left.

>👩🏻‍💻**Note:** When you build a Lex bot, you are re-training the bot with updated configurations and logic, allowing it to learn from the new parameters.

> ### Test and inspect the bot you have created so far follow the instructions in [this link](https://docs.aws.amazon.com/lexv2/latest/dg/test-bot.html).

Author: done

            
        if previous_slot_to_elicit == "text_to_translate": 
            print("diferente a none")
            text_to_translate = event["inputTranscript"]
            text_ready = TranslateText(text_to_translate,language)
            target_voice = get_target_voice(language)
            object_name,task_id = start_taskID(target_voice,bucket_name,text_ready)
            
            url_short = create_presigned_url(bucket_name, object_name, expiration=3600)
            
            print ("text_ready: ", text_ready)
            status = get_speech_synthesis(task_id)
            
            response = f"The translate text is: {text_ready}. Hear the pronunciation here {url_short} "
            messages =  [{'contentType': 'PlainText', 'content': response}]
            
            print(elicit_intent(active_contexts, session_attributes, intent, messages))
            return elicit_intent(active_contexts, session_attributes, intent, messages)
```

>🚨**Important:**  Import the necessary libraries, define the name of the bucket_name and initialize the clients for Boto3 from the AWS services, and **Deploy** to save.

### Lambda Function Permissions

> For this Lambda Function have permissions to invoke AWS services and resources, it is necessary to create an execution role that has the necessary permissions, to do this, follow the following steps: 

Author: To allow the Lambda function to invoke AWS services and resources, an execution role with the required permissions must be created. Follow these steps to create it:

1. Open the [Functions page](https://console.aws.amazon.com/lambda/home#/functions) of the Lambda console, and choose the name of a function.
2. Choose **Configuration**, and then choose **Permissions**, then Click on the **Role name** (Fig. 9).

![Slots created](images/fig_9.png)
<div align="center">Fig 9. Role name. </div> 

3. In Identity and Access Management (IAM) console, go to **Add Permision** --> **Create inline policy**
4. Select JSON, in the Policy editor copy this JSON: 

```json
{
	"Version": "2012-10-17",
	"Statement": [
		{
			"Sid": "VisualEditor0",
			"Effect": "Allow",
			"Action": [
				"polly:SynthesizeSpeech",
				"polly:StartSpeechSynthesisTask",
                "polly:GetSpeechSynthesisTask",
				"comprehend:DetectDominantLanguage",
				"translate:TranslateText"
			],
			"Resource": "*"
		},
		{
			"Sid": "VisualEditor1",
			"Effect": "Allow",
			"Action": [
				"s3:PutObject",
				"s3:GetObject"
			],
			"Resource": [
				"arn:aws:s3:::YOU-BUCKET-NAME/*",
				"arn:aws:s3:::YOU-BUCKET-NAME"
			]
		}
	]
}
```

> 🚨**Important:**  Replace **YOU-BUCKET-NAME** with your bucket name.

5. Select **Next**, write the *Policy name* and then **Create policy**.

> 🥳 Ok, you have all the code ready, now let's get it to work!

Author: done

## Part 6 - Let's get it to work!
### Attaching a Lambda function to a bot alias

> To trigger the Lambda Function when a user interacts with Amazon Lex like a Dialog code hook, Lambda Function its attaching to a bot alias [following these steps](https://docs.aws.amazon.com/lexv2/latest/dg/lambda-attach.html):

Author: To trigger a Lambda function when a user interacts with Amazon Lex, you attach the function to a bot alias by [following these steps](https://docs.aws.amazon.com/lexv2/latest/dg/lambda-attach.html)

- [Web UI for Your Chatbot](https://aws.amazon.com/blogs/machine-learning/deploy-a-web-ui-for-your-chatbot/). 

## Conclusion

Building this multilingual app using AWS services was an eye-opening learning experience. By leveraging Amazon Comprehend, Amazon Translate, Amazon Polly, and Amazon Lex, You was able to create a powerful translation tool with text-to-speech capabilities in a short time. 

The process demonstrated how easy it is to integrate AWS AI services through AWS Lambda functions. With some coding knowledge, anyone can build sophisticated applications like language translation and speech synthesis.

Experimenting hands-on is the best way to gain skills. Though translation apps already exist, creating your own solution drives learning. Building things yourself matters more than whether it's already been done.

Overall, this project exemplified how AWS cloud technology lets us rapidly turn ideas into reality. With knowledge and the right tools, we can create apps that make a difference in the world. Coding skills combined with AWS services are a powerful toolkit for building the future.

> The takeaway is if you can imagine it, you can build it with AWS. This project expanded you skills while creating something meaningful. 

Author: done
